File: Leetcode_solutions/Rotate_Image.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
const rotate = function(matrix) {
  matrix.reverse();
  const len = matrix.length;
  for (let i = 0; i < len; i++) {
    for (let j = 0; j < len; j++) {
      matrix[i].push(matrix[j][i]);
    }
  }
  for (let k = 0; k < len; k++) {
    matrix[k].splice(0, len);
  }
};

File: Leetcode_solutions/rotate_image.py
class Solution:
    def rotate(self, matrix):
        """
        Do not return anything, modify matrix in-place instead.
        """
        matrix.reverse()
        for i in range(len(matrix) - 1):
            for j in range(i, len(matrix)):
                [matrix[i][j], matrix[j][i]] = [matrix[j][i], matrix[i][j]]
        return matrix

    def rotate1(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix[0])        
        for i in range(n // 2 + n % 2):
            for j in range(n // 2):
                print(matrix[n - 1 - j][i], matrix[n - 1 - i][n - j - 1], matrix[j][n - 1 - i], matrix[i][j])
                tmp = matrix[n - 1 - j][i]
                matrix[n - 1 - j][i] = matrix[n - 1 - i][n - j - 1]
                matrix[n - 1 - i][n - j - 1] = matrix[j][n - 1 - i]
                matrix[j][n - 1 - i] = matrix[i][j]
                matrix[i][j] = tmp
